sdks/go/pkg/beam/artifact/gcsproxy/staging.go

```diff
@@ -0,0 +1,200 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcsproxy contains artifact staging and retrieval servers backed by GCS.
+package gcsproxy
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"hash"
+	"path"
+	"sync"
+
+	pb "github.com/apache/beam/sdks/go/pkg/beam/model/org_apache_beam_runner_v1"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/gcsx"
+	"github.com/golang/protobuf/proto"
+	"golang.org/x/net/context"
+	"google.golang.org/api/storage/v1"
+)
+
+// StaginServer is a artifact staging server backed by Google Cloud Storage
+// (GCS). It commits a single manifest and ignores the staging id.
+type StagingServer struct {
+	manifest     string
+	bucket, root string
+	blobs        map[string]staged
+	mu           sync.Mutex
+}
+
+type staged struct {
+	object, hash string
+}
+
+// NewStagingServer creates a artifact staging server for the given manifest.
+// It requires that the manifest is in GCS and will stage the supplied
+// artifacts next to it.
+func NewStagingServer(manifest string) (*StagingServer, error) {
+	bucket, object, err := gcsx.ParseObject(manifest)
+	if err != nil {
+		return nil, fmt.Errorf("invalid manifest location: %v", err)
+	}
+	root := path.Join(path.Dir(object), "blobs")
+
+	return &StagingServer{
+		manifest: object,
+		bucket:   bucket,
+		root:     root,
+		blobs:    make(map[string]staged),
+	}, nil
+}
+
+// CommitManifest commits the given artifact manifest to GCS.
+func (s *StagingServer) CommitManifest(ctx context.Context, req *pb.CommitManifestRequest) (*pb.CommitManifestResponse, error) {
+	manifest := req.GetManifest()
+
+	s.mu.Lock()
+	loc, err := matchLocations(manifest.GetArtifact(), s.blobs)
```

Comment:
What ifmatchLocations was a member of StagingServer so it can manage the lock, so internally it can do a 

s.mu.Lock() and defer s.mu.Unlock()

This would read a bit cleaner, and be more modular. With this approach, if someone moves/deletes line 77, chaos would ensue.

Reply:
Good point. The lock is currently held in the two public methods only, which lets matchLocations be a pure function. I think the current form is simpler than adding methods and state assumptions, but we can reevaluate if the code becomes more complicated. I wish Go had block-scoped locking, though.

Reply:
Block-scoped locking would help. That's fine, since the locking policy is cohesive and there are only a few locking methods.